casr/src/analysis.rs

```diff
@@ -0,0 +1,428 @@
+                })
+            },
+            || log_progress(&counter, total),
+        )
+        .0?;
+
+    // Deduplicate reports.
+    if output_dir.read_dir()?.count() < 2 {
+        info!("There are less than 2 CASR reports, nothing to deduplicate.");
+        return summarize_results(output_dir, crashes, gdb_argv, num_of_threads, timeout);
+    }
+    info!("Deduplicating CASR reports...");
+    let casr_cluster_d = Command::new(&casr_cluster)
+        .arg("-d")
+        .arg(output_dir.clone().into_os_string())
+        .output()
+        .with_context(|| format!("Couldn't launch {casr_cluster:?}"))?;
+
+    if casr_cluster_d.status.success() {
+        info!(
+            "{}",
+            String::from_utf8_lossy(&casr_cluster_d.stdout)
+                .lines()
+                .collect::<Vec<&str>>()
+                .join(". ")
+        );
+    } else {
+        bail!("{}", String::from_utf8_lossy(&casr_cluster_d.stderr));
+    }
+
+    if !matches.get_flag("no-cluster") {
+        if output_dir
+            .read_dir()?
+            .flatten()
+            .map(|e| e.path())
+            .filter(|e| e.extension().is_some() && e.extension().unwrap() == "casrep")
+            .count()
+            < 2
+        {
+            info!("There are less than 2 CASR reports, nothing to cluster.");
+            return summarize_results(output_dir, crashes, gdb_argv, num_of_threads, timeout);
+        }
+        info!("Clustering CASR reports...");
+        let casr_cluster_c = Command::new(&casr_cluster)
+            .arg("-c")
+            .arg(output_dir.clone().into_os_string())
+            .output()
+            .with_context(|| format!("Couldn't launch {casr_cluster:?}"))?;
+
+        if casr_cluster_c.status.success() {
+            info!(
+                "{}",
+                String::from_utf8_lossy(&casr_cluster_c.stdout).trim_end()
+            );
+        } else {
+            error!("{}", String::from_utf8_lossy(&casr_cluster_c.stderr));
+        }
+
+        // Remove reports from deduplication phase. They are in clusters now.
+        for casrep in fs::read_dir(output_dir)?.flatten().map(|e| e.path()) {
+            if let Some(ext) = casrep.extension() {
+                if ext == "casrep" {
+                    let _ = fs::remove_file(casrep);
+                }
+            }
+        }
+    }
+
+    summarize_results(output_dir, crashes, gdb_argv, num_of_threads, timeout)
+}
+
+/// Copy crashes next to reports and print summary.
+/// Run casr-gdb on uninstrumented binary if specified in ARGS.
+///
+/// # Arguments
+///
+/// * `dir` - directory with casr reports
+///
+/// * `crashes` - crashes info
+///
+/// * `gdb_args` - run casr-gdb on uninstrumented binary if specified
+///
+/// * `jobs` - number of threads for casr-gdb reports generation
+///
+/// * `timeout` - target program timeout
+fn summarize_results(
+    dir: &Path,
+    crashes: &HashMap<String, CrashInfo>,
+    gdb_args: &Vec<String>,
```

Comment:
The question is like above, could we eliminate `gdb_args`?

Reply:
done